File: Serenity.Script.UI/Widget/Flexify.cs
using jQueryApi;
using jQueryApi.UI.Interactions;
using System;
using System.Collections;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported]
    public class Flexify : Widget<FlexifyOptions>
    {
        private Int32 xDifference;
        private Int32 yDifference;

        public Flexify(jQueryObject container, FlexifyOptions options)
            : base(container, options)
        {
            var self = this;
            LazyLoadHelper.ExecuteOnceWhenShown(container, delegate {
                self.StoreInitialSize();
            });
        }
 
        private void StoreInitialSize()
        {
            if (Q.IsTrue(element.GetDataValue("flexify-init")))
                return;

            element.Data("flexify-width", options.DesignWidth ?? element.GetWidth());
            element.Data("flexify-height", options.DesignHeight ?? element.GetHeight());
            element.Data("flexify-init", true);

            var self = this;
            element.As<dynamic>().bind("resize." + this.uniqueName, new Action<jQueryEvent, dynamic>(delegate(jQueryEvent e, dynamic ui)
            {
                self.ResizeElements();
            }));

            element.As<dynamic>().bind("resizestop." + this.uniqueName, new Action<jQueryEvent, dynamic>(delegate(jQueryEvent e, dynamic ui)
            {
                self.ResizeElements();
            }));

            var tabs = element.Find(".ui-tabs");
            if (tabs.Length > 0)
                tabs.Bind("tabsactivate." + this.uniqueName, delegate
                {
                    self.ResizeElements();
                });

            if (options.DesignWidth != null ||
                options.DesignHeight != null)
                self.ResizeElements();
        }

        private Double GetXFactor(jQueryObject element)
        {
            double? xFactor = null;
            if (options.GetXFactor != null)
                xFactor = options.GetXFactor(element);

            if (xFactor == null)
                xFactor = element.GetDataValue("flex-x").As<double?>();

            return xFactor ?? 1;
        }

        private Double GetYFactor(jQueryObject element)
        {
            double? yFactor = null;
            if (options.GetYFactor != null)
                yFactor = options.GetYFactor(element);

            if (yFactor == null)
                yFactor = element.GetDataValue("flex-y").As<double?>();

            return yFactor ?? 0;
        }

        private void ResizeElements()
        {
            var width = element.GetWidth();
            var initialWidth = element.GetDataValue("flexify-width").As<int?>();
            if (initialWidth == null)
            {
                element.Data("flexify-width", width);
                initialWidth = width;
            }

            var height = element.GetHeight();
            var initialHeight = element.GetDataValue("flexify-height").As<int?>();
            if (initialHeight == null)
            {
                element.Data("flexify-height", height);
                initialHeight = height;
            }

            xDifference = width - initialWidth.Value;
            yDifference = height - initialHeight.Value;

            var self = this;

            jQueryObject containers = element;

            var tabPanels = element.Find(".ui-tabs-panel");
            if (tabPanels.Length > 0)
                containers = tabPanels.Filter(":visible");
            
            containers.Find(".flexify").Add(tabPanels.Filter(".flexify:visible")).Each((i, e) =>
            {
                self.ResizeElement(J(e));
            });
        }

        private void ResizeElement(jQueryObject element)
        {
            var xFactor = GetXFactor(element);
            if (xFactor != 0)
            {
                var initialWidth = element.GetDataValue("flexify-width").As<int?>();
                if (initialWidth == null)
                {
                    var width = element.GetWidth();
                    element.Data("flexify-width", width);
                    initialWidth = width;
                }

                element.Width(Math.Truncate(initialWidth.Value + xFactor * xDifference));
            }

            var yFactor = GetYFactor(element);
            if (yFactor != 0)
            {
                var initialHeight = element.GetDataValue("flexify-height").As<int?>();
                if (initialHeight == null)
                {
                    var height = element.GetHeight();
                    element.Data("flexify-height", height);
                    initialHeight = height;
                }

                element.Height(Math.Truncate(initialHeight.Value + yFactor * yDifference));
            }

            if (element.HasClass("require-layout"))
                element.TriggerHandler("layout");
        }
    }

    [Imported]
    public static class FlexifyExtensions
    {
        [InstanceMethodOnFirstArgument]
        public static jQueryObject FlexHeightOnly(this jQueryObject element, double flexY = 1)
        {
            return element.AddClass("flexify").Data("flex-y", flexY).Data("flex-x", 0);
        }

        [InstanceMethodOnFirstArgument]
        public static jQueryObject FlexWidthOnly(this jQueryObject element, double flexX = 1)
        {
            return element.AddClass("flexify").Data("flex-x", flexX).Data("flex-y", 0);
        }

        [InstanceMethodOnFirstArgument]
        public static jQueryObject FlexWidthHeight(this jQueryObject element, double flexX = 1, double flexY = 1)
        {
            return element.AddClass("flexify").Data("flex-x", flexX).Data("flex-y", flexY);
        }

        [InstanceMethodOnFirstArgument, ScriptName("flexX")]
        public static jQueryObject FlexXFactor(this jQueryObject element, double flexX)
        {
            return element.Data("flex-x", flexX);
        }

        [InstanceMethodOnFirstArgument, ScriptName("flexY")]
        public static jQueryObject FlexYFactor(this jQueryObject element, double flexY)
        {
            return element.Data("flex-y", flexY);
        }
    }

    [Imported, Serializable]
    public class FlexifyOptions
    {
        public Func<jQueryObject, double?> GetXFactor { get; set; }
        public Func<jQueryObject, double?> GetYFactor { get; set; }
        public Int32? DesignWidth { get; set; }
        public Int32? DesignHeight { get; set; }
    }
}

File: packages/corelib/src/ui/widgets/flexify.ts
import sQuery from "@optionaldeps/squery";
import { Widget } from "./widget";
import { LazyLoadHelper } from "../helpers/lazyloadhelper";

if (sQuery.fn) {
    sQuery.fn.flexHeightOnly = function (flexY: number = 1) {
        return this.flexWidthHeight(0, flexY);
    }

    sQuery.fn.flexWidthOnly = function (flexX: number = 1) {
        return this.flexWidthHeight(flexX, 0);
    }

    sQuery.fn.flexWidthHeight = function (flexX: number = 1, flexY: number = 1) {
        return this.addClass('flexify').data('flex-x', flexX).data('flex-y', flexY);
    }

    sQuery.fn.flexX = function (flexX: number): JQuery {
        return this.data('flex-x', flexX);
    }

    sQuery.fn.flexY = function (flexY: number): JQuery {
        return this.data('flex-y', flexY);
    }
}

export class Flexify extends Widget<FlexifyOptions> {
    private xDifference = 0;
    private yDifference = 0;
    constructor(container: JQuery, options: FlexifyOptions) {
        super({ element: container, ...options });

        LazyLoadHelper.executeOnceWhenShown(this.domNode, () => {
            this.storeInitialSize();
        });
    }

    storeInitialSize() {
        if (!!sQuery(this.domNode).data('flexify-init')) {
            return;
        }

        var designWidth = this.options.designWidth;
        if (designWidth == null)
            designWidth = sQuery(this.domNode).width();
        sQuery(this.domNode).data('flexify-width', designWidth);

        var designHeight = this.options.designHeight;
        if (designHeight == null)
            designHeight = sQuery(this.domNode).height();
        sQuery(this.domNode).data('flexify-height', designHeight);

        sQuery(this.domNode).data('flexify-init', true);

        sQuery(this.domNode).on('resize.' + this.uniqueName, () =>
            this.resizeElements());

        sQuery(this.domNode).on('resizestop.' + this.uniqueName, () =>
            this.resizeElements());
        
        var tabs = sQuery(this.domNode).find('.ui-tabs');
        if (tabs.length > 0) {
            tabs.on('tabsactivate.' + this.uniqueName, () =>
                this.resizeElements());
        }

        if (this.options.designWidth != null || this.options.designHeight != null) 
            this.resizeElements();
    }

    private getXFactor(element: JQuery): number {
        var xFactor = null;
        if (this.options.getXFactor != null)
            xFactor = this.options.getXFactor(element);

        if (xFactor == null)
            xFactor = element.data('flex-x');

        return xFactor ?? 1;
    }

    private getYFactor(element: JQuery): number {
        var yFactor = null;

        if (this.options.getYFactor != null) 
            yFactor = this.options.getYFactor(element);

        if (yFactor == null) 
            yFactor = element.data('flex-y');

        return yFactor ?? 0;
    }

    private resizeElements(): void {
        var width = sQuery(this.domNode).width();
        var initialWidth = sQuery(this.domNode).data('flexify-width');
        if (initialWidth == null) {
            sQuery(this.domNode).data('flexify-width', width);
            initialWidth = width;
        }

        var height = sQuery(this.domNode).height();
        var initialHeight = sQuery(this.domNode).data('flexify-height');
        if (initialHeight == null) {
            sQuery(this.domNode).data('flexify-height', height);
            initialHeight = height;
        }

        this.xDifference = width - initialWidth;
        this.yDifference = height - initialHeight;

        var containers = sQuery(this.domNode);
        var tabPanels = sQuery(this.domNode).find('.ui-tabs-panel');

        if (tabPanels.length > 0) {
            containers = tabPanels.filter(':visible');
        }

        containers.find('.flexify')
            .add(tabPanels.filter('.flexify:visible'))
            .each((i, e) => {
                this.resizeElement(sQuery(e as any));
            });
    }

    private resizeElement(element: JQuery): void {
        var xFactor = this.getXFactor(element);
        if(xFactor !== 0) {
            var initialWidth = element.data('flexify-width');
            if (initialWidth == null) {
                var width = element.width();
                element.data('flexify-width', width);
                initialWidth = width;
            }

            element.width(initialWidth + xFactor * this.xDifference | 0);
        }

        var yFactor = this.getYFactor(element);
        if (yFactor !== 0) {
            var initialHeight = element.data('flexify-height');
            if (initialHeight == null) {
                var height = element.height();
                element.data('flexify-height', height);
                initialHeight = height;
            }
            element.height(initialHeight + yFactor * this.yDifference | 0);
        }

        if (element.hasClass('require-layout')) {
            element.triggerHandler('layout');
        }
    }
}

export interface FlexifyOptions {
    getXFactor?: (p1: JQuery) => any;
    getYFactor?: (p1: JQuery) => any;
    designWidth?: any;
    designHeight?: any;
}